File: GameLibrary/src/cc/lib/zombicide/ZUser.java
package cc.lib.zombicide;

import java.util.ArrayList;
import java.util.List;

import cc.lib.game.GColor;
import cc.lib.game.Utils;

public abstract class ZUser {

    public final static GColor [] USER_COLORS = {
            GColor.YELLOW,
            GColor.RED,
            GColor.GREEN,
            GColor.ORANGE,
            GColor.BLUE,
            GColor.MAGENTA
    };

    public final static String [] USER_COLOR_NAMES = {
            "YELLOW", "RED", "GREEN", "ORANGE", "BLUE", "MAGENTA"
    };

    private int color = 0;
    private String name = null;
    private final List<ZPlayerName> characters = new ArrayList<>();

    public GColor getColor() {
        return USER_COLORS[color];
    }

    public int getColorId() {
        return color;
    }

    public void setColor(int color) {
        Utils.assertTrue(color >= 0 && color < USER_COLORS.length);
        this.color = color;
        for (ZPlayerName nm : characters) {
            nm.character.setColor(USER_COLORS[color]);
        }
    }

    public String getName() {
        return name == null ? USER_COLOR_NAMES[color] : name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<ZPlayerName> getCharacters() {
        return new ArrayList<>(characters);
    }

    public void clearCharacters() {
        characters.clear();
    }

    public void addCharacter(ZPlayerName c) {
        characters.remove(c);
        characters.add(c);
        c.character.setColor(USER_COLORS[color]);
    }

    public void removeCharacter(ZPlayerName name) {
        name.character.setColor(null);
        characters.remove(name);
    }

    public void setCharacters(List<ZPlayerName> chars) {
        characters.clear();
        for (ZPlayerName nm : chars) {
            addCharacter(nm);
        }
    }

    public void setCharactersHidden(boolean hidden) {
        for (ZPlayerName nm : characters) {
            if (nm.getCharacter() != null)
                nm.getCharacter().setInvisible(hidden);
        }
    }

    public abstract ZPlayerName chooseCharacter(List<ZPlayerName> options);

    public abstract Integer chooseMove(ZPlayerName cur, List<ZMove> options);

    <T> T elemOrNull(Integer idx, List options) {
        if (idx == null)
            return null;
        return (T)options.get(idx);
    }

    final ZMove chooseMoveInternal(ZPlayerName cur, List<ZMove> options) {
        return elemOrNull(chooseMove(cur, options), options);
    }

    public abstract ZSkill chooseNewSkill(ZPlayerName character, List<ZSkill> skillOptions);

    public abstract ZEquipSlot chooseSlotToOrganize(ZPlayerName cur, List<ZEquipSlot> slots);

    public abstract Integer chooseEquipment(ZPlayerName cur, List<ZEquipment> equipOptions);

    final ZEquipment chooseEquipmentInternal(ZPlayerName cur, List<ZEquipment> equipOptions) {
        return elemOrNull(chooseEquipment(cur, equipOptions), equipOptions);
    }

    public abstract ZEquipSlot chooseSlotForEquip(ZPlayerName cur, List<ZEquipSlot> equipableSlots);

    public abstract Integer chooseZoneToWalk(ZPlayerName cur, List<Integer> zones);

    public abstract Integer chooseDoorToToggle(ZPlayerName cur, List<ZDoor> doors);

    final ZDoor chooseDoorToToggleInternal(ZPlayerName cur, List<ZDoor> doors) {
        return elemOrNull(chooseDoorToToggle(cur, doors), doors);
    }

    public abstract Integer chooseWeaponSlot(ZPlayerName c, List<ZWeapon> weapons);

    final ZWeapon chooseWeaponSlotInternal(ZPlayerName c, List<ZWeapon> weapons) {
        return elemOrNull(chooseWeaponSlot(c, weapons), weapons);
    }

    public abstract ZPlayerName chooseTradeCharacter(ZPlayerName c, List<ZPlayerName> list);

    public abstract Integer chooseZoneForAttack(ZPlayerName c, List<Integer> zones);

    public abstract Integer chooseItemToPickup(ZPlayerName cur, List<ZEquipment> list);

    final ZEquipment chooseItemToPickupInternal(ZPlayerName cur, List<ZEquipment> list) {
        return elemOrNull(chooseItemToPickup(cur, list), list);
    }

    public abstract Integer chooseItemToDrop(ZPlayerName cur, List<ZEquipment> list);

    final ZEquipment chooseItemToDropInternal(ZPlayerName cur, List<ZEquipment> list) {
        return elemOrNull(chooseItemToDrop(cur, list), list);
    }

    public abstract Integer chooseEquipmentToThrow(ZPlayerName cur, List<ZEquipment> slots);

    final ZEquipment chooseEquipmentToThrowInternal(ZPlayerName cur, List<ZEquipment> slots) {
        return elemOrNull(chooseEquipmentToThrow(cur, slots), slots);
    }

    public abstract Integer chooseZoneToThrowEquipment(ZPlayerName cur, ZEquipment toThrow, List<Integer> zones);

    public abstract Integer chooseZoneToShove(ZPlayerName cur, List<Integer> list);

    public abstract ZSpell chooseSpell(ZPlayerName cur, List<ZSpell> spells);

    public abstract ZPlayerName chooseCharacterForSpell(ZPlayerName cur, ZSpell spell, List<ZPlayerName> targets);

    public abstract ZPlayerName chooseCharacterToBequeathMove(ZPlayerName cur, List<ZPlayerName> list);

    public abstract Integer chooseZoneForBloodlust(ZPlayerName cur, List<Integer> list);

    public abstract Integer chooseSpawnAreaToRemove(ZPlayerName cur, List<ZSpawnArea> list);

    public abstract Integer chooseZoneToIgnite(ZPlayerName playerName, List<Integer> ignitableZones);
}

File: Zombicide/src/cc/lib/zombicide/ZUser.kt
package cc.lib.zombicide

import cc.lib.game.GColor
import cc.lib.ksp.remote.IRemoteSuspend
import cc.lib.ksp.remote.Remote
import cc.lib.ksp.remote.RemoteFunction

@Remote
abstract class ZUser(_name: String?, var colorId: Int) : IRemoteSuspend {

	init {
		if (colorId !in USER_COLORS.indices)
			throw IllegalArgumentException("colorId out of bounds")
	}

	var name = _name ?: USER_COLOR_NAMES[colorId]

	private val _players = mutableSetOf<ZPlayerName>()

	val players: List<ZPlayerName>
		get() = _players.toList()

	fun clearCharacters() {
		_players.clear()
	}

	fun addCharacter(c: ZCharacter) {
		_players.add(c.type)
		c.colorId = colorId
		c.isInvisible = false
	}

	fun removeCharacter(character: ZCharacter) {
		_players.remove(character.type)
		character.colorId = -1
	}

	fun setCharacters(chars: List<ZCharacter>) {
		_players.clear()
		for (nm in chars) {
			addCharacter(nm)
		}
	}

	fun getColor(): GColor = USER_COLORS[colorId]

	fun getColorName(): String = USER_COLOR_NAMES[colorId]

	fun hasPlayer(name: ZPlayerName): Boolean {
		return _players.contains(name)
	}

	@RemoteFunction
	abstract suspend fun chooseCharacter(options: List<ZPlayerName>): ZPlayerName?

	@RemoteFunction
	abstract suspend fun chooseMove(cur: ZPlayerName, options: List<ZMove>): ZMove?

	@RemoteFunction
	abstract suspend fun chooseNewSkill(character: ZPlayerName, skillOptions: List<ZSkill>): ZSkill?

	@RemoteFunction
	abstract suspend fun chooseSlotToOrganize(cur: ZPlayerName, slots: List<ZEquipSlot>): ZEquipSlot?

	@RemoteFunction
	abstract suspend fun chooseEquipment(cur: ZPlayerName, equipOptions: List<ZEquipment<*>>): Int?
	suspend fun chooseEquipmentInternal(cur: ZPlayerName, equipOptions: List<ZEquipment<*>>): ZEquipment<*>? {
		return chooseEquipment(cur, equipOptions)?.let { equipOptions[it] }
	}

	@RemoteFunction
	abstract suspend fun chooseSlotForEquip(cur: ZPlayerName, equipableSlots: List<ZEquipSlot>): ZEquipSlot?

	@RemoteFunction
	abstract suspend fun chooseZoneToWalk(cur: ZPlayerName, zones: List<Int>): Int?

	@RemoteFunction
	abstract suspend fun chooseDoorToToggle(cur: ZPlayerName, doors: List<ZDoor>): Int?

	suspend fun chooseDoorToToggleInternal(cur: ZPlayerName, doors: List<ZDoor>): ZDoor? {
		return chooseDoorToToggle(cur, doors)?.let { doors[it] }
	}

	@RemoteFunction
	abstract suspend fun chooseWeaponSlot(c: ZPlayerName, weapons: List<ZWeapon>): Int?
	suspend fun chooseWeaponSlotInternal(c: ZPlayerName, weapons: List<ZWeapon>): ZWeapon? {
		return chooseWeaponSlot(c, weapons)?.let { weapons[it] }
	}

	@RemoteFunction
	abstract suspend fun chooseTradeCharacter(c: ZPlayerName, list: List<ZPlayerName>): ZPlayerName?

	@RemoteFunction
	abstract suspend fun chooseZoneForAttack(c: ZPlayerName, zones: List<Int>): Int?

	@RemoteFunction
	abstract suspend fun chooseItemToPickup(cur: ZPlayerName, list: List<ZEquipment<*>>): Int?
	suspend fun chooseItemToPickupInternal(cur: ZPlayerName, list: List<ZEquipment<*>>): ZEquipment<*>? {
		return chooseItemToPickup(cur, list)?.let { list[it] }
	}

	@RemoteFunction
	abstract suspend fun chooseItemToDrop(cur: ZPlayerName, list: List<ZEquipment<*>>): Int?

	suspend fun chooseItemToDropInternal(cur: ZPlayerName, list: List<ZEquipment<*>>): ZEquipment<*>? {
		return chooseItemToDrop(cur, list)?.let { list[it] }
	}

	@RemoteFunction
	abstract suspend fun chooseEquipmentToThrow(cur: ZPlayerName, slots: List<ZEquipment<*>>): Int?
	suspend fun chooseEquipmentToThrowInternal(
		cur: ZPlayerName,
		slots: List<ZEquipment<*>>
	): ZEquipment<*>? {
		return chooseEquipmentToThrow(cur, slots)?.let { slots[it] }
	}

	@RemoteFunction
	abstract suspend fun chooseZoneToThrowEquipment(
		cur: ZPlayerName,
		toThrow: ZEquipment<*>,
		zones: List<Int>
	): Int?

	@RemoteFunction
	abstract suspend fun chooseZoneToShove(cur: ZPlayerName, list: List<Int>): Int?

	@RemoteFunction
	abstract suspend fun chooseSpell(cur: ZPlayerName, spells: List<ZSpell>): ZSpell?

	@RemoteFunction
	abstract suspend fun chooseCharacterForSpell(
		cur: ZPlayerName,
		spell: ZSpell,
		targets: List<ZPlayerName>
	): ZPlayerName?

	@RemoteFunction
	abstract suspend fun chooseCharacterToBequeathMove(
		cur: ZPlayerName,
		list: List<ZPlayerName>
	): ZPlayerName?

	@RemoteFunction
	abstract suspend fun chooseZoneForBloodlust(cur: ZPlayerName, list: List<Int>): Int?

	@RemoteFunction
	abstract suspend fun chooseSpawnAreaToRemove(cur: ZPlayerName, list: List<ZSpawnArea>): Int?

	@RemoteFunction
	abstract suspend fun chooseZoneToIgnite(
		playerName: ZPlayerName,
		ignitableZones: List<Int>
	): Int?

	@RemoteFunction
	abstract suspend fun chooseEquipmentClass(
		playerName: ZPlayerName,
		classes: List<ZEquipmentClass>
	): ZEquipmentClass?

	@RemoteFunction
	abstract suspend fun chooseStartingEquipment(
		playerName: ZPlayerName,
		list: List<ZEquipmentType>
	): ZEquipmentType?

	@RemoteFunction
	abstract suspend fun chooseFamiliar(
		playerName: ZPlayerName,
		list: List<ZFamiliarType>
	): ZFamiliarType?

	@RemoteFunction
	abstract suspend fun chooseOrganize(playerName: ZPlayerName, list: List<ZMove>, undos: Int): ZMove?

	@RemoteFunction
	abstract suspend fun chooseZoneForCatapult(
		playerName: ZPlayerName,
		ammoType: ZWeaponType,
		zones: List<Int>
	): Int?

	/**
	 * This may be called more than once. Need to handle if a dialog is already open
	 */
	@RemoteFunction
	open suspend fun organizeStart(primary: ZPlayerName, secondary: ZPlayerName?, undos: Int) {
		throw NotImplementedError()
	}

	@RemoteFunction
	open suspend fun organizeEnd() {
		throw NotImplementedError()
	}

	companion object {

		@JvmField
		val USER_COLORS = arrayOf(
			GColor.YELLOW,
			GColor.RED,
			GColor.GREEN,
			GColor.ORANGE,
			GColor.BLUE,
			GColor.MAGENTA
		)

		@JvmField
		val USER_COLOR_NAMES = arrayOf(
			"YELLOW", "RED", "GREEN", "ORANGE", "BLUE", "MAGENTA"
		)

		fun getColorName(id: Int): String = if (id in USER_COLORS.indices) {
			USER_COLOR_NAMES[id]
		} else "None($id)"

		fun getColorName(color: GColor): String = USER_COLORS.indexOfFirst { it == color }.takeIf { it >= 0 }?.let {
			USER_COLOR_NAMES[it]
		} ?: "Unknown($color)"
	}
}